conn.go

```diff
@@ -493,27 +532,20 @@ func (l *Conn) processMessages() {
 
 				// Add timeout if defined
 				if l.requestTimeout > 0 {
-					go func() {
-						timer := time.NewTimer(time.Duration(l.requestTimeout))
-						defer func() {
-							if err := recover(); err != nil {
-								logger.Printf("ldap: recovered panic in RequestTimeout: %v", err)
-							}
-
-							timer.Stop()
-						}()
-
-						select {
-						case <-timer.C:
-							timeoutMessage := &messagePacket{
-								Op:        MessageTimeout,
-								MessageID: message.MessageID,
-							}
-							l.sendProcessMessage(timeoutMessage)
-						case <-message.Context.done:
-						}
-					}()
+					message.Context.ctx, _ = context.WithTimeout(message.Context.ctx, time.Duration(l.requestTimeout))
```

Comment:
is there risk of the context and the stored cancelation func getting out of sync here?

Reply:
No - because it's inherited from the prior context, and cancels cascade. When the parent is cancelled, it'll take them both out and tear down everything.